pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -722,6 +735,24 @@ protected WriteInEventLoopCallback newObject(Handle<WriteInEventLoopCallback> ha
 
     @Override
     public CompletableFuture<Void> closeAsync() {
+        return closeAsync(true);
+    }
+
+    public CompletableFuture<Void> closeAsync(boolean waitForPendingMessage) {
+        CompletableFuture<Void> closeResult = new CompletableFuture<>();
+        closeAsync(waitForPendingMessage, closeResult);
+        return closeResult;
+    }
+
+    private void closeAsync(boolean waitForPendingMessage, CompletableFuture<Void> closeFuture) {
+        if (waitForPendingMessage && !pendingMessages.isEmpty()) {
+            CLOSING_IN_PROGRESS_UPDATER.set(this, TRUE);
```

Comment:
If we're not using any `compareAndSet()` over the flag, then we won't need to use the updater.